<link rel="stylesheet" href="estilo.css">

<img src="img/logoifsc.png">

---

Prof. Carlos Andrés Ferrero

# Mineração de Dados

## Pré-Processamento de Dados
    
---

# Conteúdo

Pré-processamento de dados quantiativos e qualitativos

- Limpeza de dados
- Transformação de dados
    - Normalização de dados numéricos
    - Discretização de dados numéricos
    - Transformação de dados qualitativos em quantiativos
    - Outras transformações de dados
- Tratamento de dados desconhecidos

# Limpeza de dados

- A maior parte dos conjuntos de dados disponíveis para estudo, bem como os obtidos a partir de processos de coleta de dados, podem apresentar problemas, os quais dificultam a utilização desses dados em processos de mineração.

- Esses problemas pode ser: erros de digitação, erro na leitura de dados por sensores.

- Muitos dos processos de mineração aplicados não são usados apenas para estudo, mas os padrões encontrados podem ser usados em processos de mineração de dados. Por esse fato, é extremente importante a limpeza desses dados.

## Baseada em Conhecimento do domínio
- Uma estratégia para fazer limpeza de dados é **utilizar conhecimento do domínio**
- Determinados valores numéricos podem estar fora de um intervalo definido, conhecido do domínio.

- Um atributo idade deve conter valores entre 0 e 120 para domínio do problema. Se algum valor desse atributo está fora desse intervalo, temos que tomar uma decisão sobre ele: deixar ele como desconhecido (NaN no numpy), escolher algum valor para ele, ou tentar encontrar o valor verdadeiro baseado em outros atributos ou instâncias.

In [2]:
import numpy as np
idades = np.array([10,20,34,-25,50,165,71,55,30,34], dtype= np.float32)

In [3]:
desconhecido = np.NaN
idades = np.where( np.logical_or(idades < 0, idades > 120), desconhecido, idades)

In [4]:
idades

array([10., 20., 34., nan, 50., nan, 71., 55., 30., 34.], dtype=float32)

- Um atributo de data deve conter valores de data maiores do que 01/01/2010. Se algum valor desse atributo for menor do que essa data também tamos que tomar uma decisão.

## Baseado em dados estatísticos
- Outra estratégia consiste em **utilizar estatística** para conhecer a distribuição dos dados e assim tomar decisões sobre quais valores estão fora do intervalo esperado.
- Uma variável contínua pode apresentar outliers estatísticos (pontos fora da curva) e se essa variável tem uma distribuição conhecida podemos tomar alguma decisão sobre o que fazer com esses valores: como deixar ele como desconhecido ou substituído pelo valor esperado (média na distribuição normal).

- Por exemplo, um atributo de temperatura possui um valor de média de 26 graus com desvio padrão de 2 graus. 
- Considerando uma distribuição normal espera-se que $68.25\%$ estejam a uma unidade de desvio-padrão e $99.73\%$ a três unidades.
- Se encontrar um valor acima de unidades de desvio padrão, por exemplo, 32 graus, podemos tomar alguma decisão com esse valor.

In [5]:
temperatura = np.array([26.6, 27.8, 25.6, 26.2, 120.0, 20.4, 29.4, 24.4, 26.9, 26.7])
media = 26
desvio = 2

In [6]:
temperatura = np.where( ( np.abs(temperatura - media)/desvio ) > 3, media, temperatura )

In [7]:
temperatura

array([26.6, 27.8, 25.6, 26.2, 26. , 20.4, 29.4, 24.4, 26.9, 26.7])

# Transformação de Dados

## Normalização de Dados Numéricos (ou Contínuos)

- A motivação da normalização é minimzar o problema da presença de variáveis em um conjunto de dados com dispersões de dados muito diferentes.

- Vários algoritmos de aprendizado, como Redes Neurais, Vizinhos mais Próximos e K-médias podem ser prejudicados por essa dispersão. 

- Por exemplo, se medimos a distância entre dois usários pela diferença entre seus valores temos:
    - Usuário 1 (salario = 1000, poupança = 20000)
    - Usuário 2 (salario = 1500, poupança = 50000)
    - Diferença  (salario = 500, poupança = 30000), Total (30500)
    - O problema que observamos é que a variável salário contribui muito menos para a soma das distâncias do que a variável poupança. 

- O objetivo da normalização é ajustar a escala dos valores dos atributos para que tenha dispersões ou intervalos mais similares.
- Métodos para normalização de dados numéricos:
    - Linear
    - Desvio-padrão
    - Valor máximo
    - Escala decimal

### Normalização Linear

- A normalização linear consiste em manter os dados no intervalo $[0,1]$. Para isso, uma função de transformação linear.

$$ f(x) = \frac{x - min(\mathbf{x})}{max(\mathbf{x}) - min(\mathbf{x})} = x'$$

- onde,
    - $\mathbf{x}$ é o conjunto de valores da variável
    - $x$ é um valor da variável
    - $x'$ é o valor da variável transformado

### Normalização por Desvio-Padrão (z-score)

- A normalização desvio-padrão consiste em transformar os dados em unidades de desvio padrão em relação à média. Para isso, uma função de transformação baseada em média e desvio padrão.

$$ f(x) = \frac{x - mean(\mathbf{x})}{sd(\mathbf{x})} = x'$$

- onde, 
    - $\mathbf{x}$ é o conjunto de valores da variável
    - $x$ é um valor da variável
    - $x'$ é o valor da variável transformado
    - $mean$ e $sd$ são a média e o desvio padrão dos valores

### Normalização de Valor Máximo

- A normalização usando o valor máximo consiste em manter os dados no intervalo $[-1,1]$. Para isso, uma função de transformação

$$ f(x) = \frac{x}{max(|\mathbf{x}|)} = x'$$

- onde, 
    - $\mathbf{x}$ é o conjunto de valores da variável
    - $x$ é um valor da variável
    - $x'$ é o valor da variável transformado

### Normalização por Escala Decimal

- A normalização usando o valor por escala decimal consiste em manter os dados no intervalo $[-1,1]$. Para isso, uma função de transformação

$$ f(x) = \frac{x}{10^j} = x'$$

- onde, 
    - $\mathbf{x}$ é o conjunto de valores da variável
    - $x$ é um valor da variável
    - $x'$ é o valor da variável transformado
    - $j$ é o menor inteiro tal que $max(|\mathbf{x}|) \leq 10^j$

### Exemplo no conjunto de dados Iris

In [40]:
import pandas as pd
from sklearn import datasets
iris = datasets.load_iris()
data=pd.DataFrame(iris['data'],columns=iris['feature_names'])
data

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

- Estatísticas Descritivas

In [44]:
data.describe().round(5)

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count          150.00000         150.00000           150.0000   
mean             5.84333           3.05733             3.7580   
std              0.82807           0.43587             1.7653   
min              4.30000           2.00000             1.0000   
25%              5.10000           2.80000             1.6000   
50%              5.80000           3.00000             4.3500   
75%              6.40000           3.30000             5.1000   
max              7.90000           4.40000             6.9000   

       petal width (cm)  
count         150.00000  
mean            1.19933  
std             0.76224  
min             0.10000  
25%             0.30000  
50%             1.30000  
75%             1.80000  
max             2.50000

- Definição de uma função para normalização linear

In [10]:
def normalizacao_linear(x):    
    return (x - np.min(x)) / (np.max(x) - np.min(x));

- Aplicação da função em uma das colunas do conjunto de dados

In [48]:
new_data = normalizacao_linear( data['sepal length (cm)'] )
new_data.head()

0    0.222222
1    0.166667
2    0.111111
3    0.083333
4    0.194444
Name: sepal length (cm), dtype: float64

- Aplicação da função em todas as colunas do conjunto de dados

In [51]:
new_data = data.apply(normalizacao_linear, axis = 0)
new_data

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0             0.222222          0.625000           0.067797          0.041667
1             0.166667          0.416667           0.067797          0.041667
2             0.111111          0.500000           0.050847          0.041667
3             0.083333          0.458333           0.084746          0.041667
4             0.194444          0.666667           0.067797          0.041667
..                 ...               ...                ...               ...
145           0.666667          0.416667           0.711864          0.916667
146           0.555556          0.208333           0.677966          0.750000
147           0.611111          0.416667           0.711864          0.791667
148           0.527778          0.583333           0.745763          0.916667
149           0.444444          0.416667           0.694915          0.708333

[150 rows x 4 columns]

In [13]:
new_data.describe().round(5)

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count          150.00000         150.00000          150.00000   
mean             0.42870           0.44056            0.46746   
std              0.23002           0.18161            0.29920   
min              0.00000           0.00000            0.00000   
25%              0.22222           0.33333            0.10169   
50%              0.41667           0.41667            0.56780   
75%              0.58333           0.54167            0.69492   
max              1.00000           1.00000            1.00000   

       petal width (cm)  
count         150.00000  
mean            0.45806  
std             0.31760  
min             0.00000  
25%             0.08333  
50%             0.50000  
75%             0.70833  
max             1.00000

### Atividade

- Com o conjunto de dados *iris*, para cada atributo do conjunto de dados faça:
    - Normalização Linear
    - Normalização por Desvio-padrão
    - Normalização por Valor Máximo
    - Normalização por Escala Decimal

## Discretização de Dados Numéricos

- Alguns algoritmos de aprendizado trabalham apenas com dados discretos (ou categóricos), como o ID3 (Iterative Dichotomiser 3) desenvolvido por Ross Quinlan em 1986. Isto é, ao invés de utilizar o valor de, por exemplo, altura em metros ou centímetros, utilizam os valores baixo, médio, alto.

- Outros algoritmos podem ter desempenho melhor em alguns problemas se usada a representação discretazada de uma variável contínua, como no caso do algoritmo Naïve Bayes.

- Em algumas ocasiões o armazenamento de dados é feito com valores numéricos, embora indiquem informações categóricas, por exemplo um atributo gênero, onde o valor 0 indica masculino, 1 feminino e 2 outro.

- O objetivo da categorização consiste em transformar de valores numéricos para categóricos ou discretos.

- A idéia é perder a menor quantidade de informação possível para o domínio do problema, durante esse processo.

- As principais abordagens são:
    - Mapeamento Direto
    - Mapeamento em Intervalos

### Mapeamento Direto

- Neste tipo de discretização ocorre a substituição de valores numéricos por valores categóricos conhecidos no domínio do problema.

- Como o exemplo citado anteriormente, o atributo gênero pode ser armazenado como um valor numérico e o mesmo deve ser transformado, com o objetivo que, tanto as técnicas visualização, quanto os algoritmos de aprendizado não entendam esse atributo como numérico e sim como discreto.

- Vamos criar um vetor de dados para uma variável gênero, distribuída de forma uniforme entre os valores 0, 1 e 2.

In [29]:
np.random.seed(1)
genero =  np.random.randint(low = 0, high = 3, size=20)
genero

array([1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 2, 1, 2, 0, 2, 1, 2, 0, 0])

- Criamos um DataFrame a partir dessa coluna

In [30]:
df = pd.DataFrame( {'genero' : genero} )
df

genero
0        1
1        0
2        0
3        1
4        1
5        0
6        0
7        1
8        0
9        1
10       0
11       2
12       1
13       2
14       0
15       2
16       1
17       2
18       0
19       0

- Vamos acessar a coluna `genero` e executar a função `replace` que recebe as substituições como parâmetro.

In [31]:
df['genero_categoria'] = df.genero.replace(to_replace = {0 : 'Masculino', 1 : 'Feminino', 2 : 'Outro'})
df

genero genero_categoria
0        1         Feminino
1        0        Masculino
2        0        Masculino
3        1         Feminino
4        1         Feminino
5        0        Masculino
6        0        Masculino
7        1         Feminino
8        0        Masculino
9        1         Feminino
10       0        Masculino
11       2            Outro
12       1         Feminino
13       2            Outro
14       0        Masculino
15       2            Outro
16       1         Feminino
17       2            Outro
18       0        Masculino
19       0        Masculino

- Em geral cria-se uma variável que armazena essa informações. Essa variável chamada de dicionário (tipo dict).

In [32]:
to_replace = {
    0 : 'Masculino',
    1 : 'Feminino',
    2 : 'Outro'
}
print(to_replace)
print('type: ', type(to_replace) )

{0: 'Masculino', 1: 'Feminino', 2: 'Outro'}
type:  <class 'dict'>


- Podemos passar essa variável `to_replace` como argumento do parâmetro `to_replace` da função `replace`.

In [43]:
df['genero_categoria'] = df.genero.replace(to_replace = to_replace)
df

genero genero_categoria
0        1         Feminino
1        0        Masculino
2        0        Masculino
3        1         Feminino
4        1         Feminino
5        0        Masculino
6        0        Masculino
7        1         Feminino
8        0        Masculino
9        1         Feminino
10       0        Masculino
11       2            Outro
12       1         Feminino
13       2            Outro
14       0        Masculino
15       2            Outro
16       1         Feminino
17       2            Outro
18       0        Masculino
19       0        Masculino

- Uma segunda forma de fazer isso é definir uma função `get_categoria_genero` que é responsável por receber um valor da variável (0, 1 ou 2) e retorna o valor correspondente. Duas implementações dessa função são apresentadas a seguir:

In [27]:
def get_categoria_genero(x) :
    if (x == 0) :
        return 'Masculino'
    elif (x == 1) :
        return 'Feminino'
    elif (x == 2) :
        return 'Outro'
    
get_categoria_genero(1)

'Feminino'

In [26]:
def get_categoria_genero(x) :
    dict_genero = {0 : 'Masculino', 1 : 'Feminino', 2 : 'Outro'}
    return dict_genero[x]

get_categoria_genero(0)

'Masculino'

- Aplicamos essa função na coluna `genero` armazenamos o resultado na coluna `genero_categoria_2`.

In [28]:
df['genero_categoria_2'] = df.genero.apply(get_categoria_genero)
df

genero genero_categoria genero_categoria_2
0        1         Feminino           Feminino
1        0        Masculino          Masculino
2        0        Masculino          Masculino
3        1         Feminino           Feminino
4        1         Feminino           Feminino
5        0        Masculino          Masculino
6        0        Masculino          Masculino
7        1         Feminino           Feminino
8        0        Masculino          Masculino
9        1         Feminino           Feminino
10       0        Masculino          Masculino
11       2            Outro              Outro
12       1         Feminino           Feminino
13       2            Outro              Outro
14       0        Masculino          Masculino
15       2            Outro              Outro
16       1         Feminino           Feminino
17       2            Outro              Outro
18       0        Masculino          Masculino
19       0        Masculino          Masculino

### Mapeamento em Intervalos

- Intervalos pré-definidos
- Intervalos de igual tamanho
- Intervalos de igual frequência

#### Intervalos Pré-definidos

- Neste tipo de intervalo, coleta-se informação do domínio para poder realizar a discretização

- Por exemplo, ao discretizar valores de idade em faixas etárias, embora possamos definir os intervalos arbitrariamente (nós mesmos) existe conhecimento do domínio que permite definir esses intervalos.
    - Intervalo 0 (de 0 a 12 anos incompletos)
    - Intervalo 1 (de 12 anos a 18 anos incompletos)
    - Intervalo 2 (de 18 anos a 40 anos incompletos)
    - Intervalo 3 (de 40 anos a 60 anos incompletos)
    - Intervalo 4 (de 60 anos ou mais)

In [45]:
np.random.seed(1)
idade =  np.random.randint(low = 0, high = 100, size=20)
df = pd.DataFrame({'idade' : idade})
df

idade
0      37
1      12
2      72
3       9
4      75
5       5
6      79
7      64
8      16
9       1
10     76
11     71
12      6
13     25
14     50
15     20
16     18
17     84
18     11
19     28

In [46]:
def get_discretized_idade (x) :
    if (x < 12) :
        return 0
    elif (x < 18) :
        return 1
    elif (x < 40) :
        return 2
    elif (x < 60) :
        return 3
    else :
        return 4

In [47]:
df['idade_discretizado'] = df.idade.apply(get_discretized_idade)
df

idade  idade_discretizado
0      37                   2
1      12                   1
2      72                   4
3       9                   0
4      75                   4
5       5                   0
6      79                   4
7      64                   4
8      16                   1
9       1                   0
10     76                   4
11     71                   4
12      6                   0
13     25                   2
14     50                   3
15     20                   2
16     18                   2
17     84                   4
18     11                   0
19     28                   2

### Intervalos de Igual Tamanho (ou largura fixa)

- Ao se conhecer os valores máximos e mínimos do domínio de uma variável, seja por conhecimento do domínio ou pela aplicação de técnicas estatísticas, podemos dividir o intervalo total dos dados em $k$ tamanhos iguais.

$$L = \frac{max(\mathbf{x}) - min(\mathbf{x}) + 1}{k}$$

- onde, 
    - $\mathbf{x}$ é o conjunto de valores da variável    
    - $k$ é o número de intervalos
    - $L$ é a largura do intervalo

- Por exemplo, se os valores vão de 1 a 100 e queremos dividir esses vem valores em 4 intervalos de tamanho igual, o tamanho do intervalo é 25, pois $\frac{(1-100+1)}{4} = 25$

- Assim, os intervalos podem ser definidos da seguinte por,

$$ I_i = [(min(\mathbf{x}) + L \times i),\;(min(\mathbf{x}) + L \times (i+1) - 1],\\ \text{para}\;1 \leq i < k $$

- O que resulta nos seguintes intervalos de 1 a 100

$$ 
 \begin{matrix}
  I_0 =& [(1 + 25 \times 0),(1 + 25 \times (0+1) - 1)]) =& [1,25] \\
  I_1 =& [(1 + 25 \times 1),(1 + 25 \times (1+1) - 1)]) =& [26,50] \\
  I_2 =& [(1 + 25 \times 2),(1 + 25 \times (2+1) - 1)]) =& [51,75] \\
  I_3 =& [(1 + 25 \times 3),(1 + 25 \times (3+1) - 1)]) =& [76,100] \\
  \end{matrix}
$$

- Vamos construir um conjunto de dados de 500 valores de idade entre 1 e 100.

In [52]:
np.random.seed(1)
idade =  np.random.randint(low = 1, high = 101, size=500)
print(idade)

[ 38  13  73  10  76   6  80  65  17   2  77  72   7  26  51  21  19  85
  12  29  30  15  51  69  88  88  95  97  87  14  10   8  64  62  23  58
   2   1  61  82   9  89  14  48  73  31  72   4  71  22  50  58   4  69
  25  44  77  27  53  81  42  83  16  65  69  26  99  88   8  27  26  23
  10  68  24  28  38  58  84  39   9  33  35  11  24  16  88  26  72  93
  75  63  47  33  89  24  56  66  78   4   1  78   7  53  86  71   3  77
  92  22  76   8  78  73  76  77  44  21  31  37   8  46  69  58  83  97
  14  11  24  82   8  25  75  93  21  33  13  66  95  61  25  83  98   3
  93  99  11  55  97  83  87  71  67  72  49  55  16   6  18  43  21  49
  23  14  98  54  85  11  97  56  62  57  90  22  97  84  26  15  14  85
  44   7  78  57  60  16  25  10  67  72  54  70  37  22  41  78  92  50
  48  78  41  79  46  88  17  29  46  68  67  79  47   1  30  64  76  36
  54  94  34   3  85  84  49  55  33  29  56  83  32  29  95  75   9 100
  33   9  85  78  51  80  42  65  84  25  21  45  1

- Calculamos a largura do intervalo $L$ conforme a equação:

$$L = \frac{max(\mathbf{x}) - min(\mathbf{x}) + 1}{k}$$

In [77]:
k = 4
L = (max(idade) - min(idade) + 1) / k
print(L)

25.0


In [78]:
intervalos = (idade - min(idade)) // L
intervalos_largura_fixa = intervalos.astype(int)
intervalos_largura_fixa

array([1, 0, 2, 0, 3, 0, 3, 2, 0, 0, 3, 2, 0, 1, 2, 0, 0, 3, 0, 1, 1, 0,
       2, 2, 3, 3, 3, 3, 3, 0, 0, 0, 2, 2, 0, 2, 0, 0, 2, 3, 0, 3, 0, 1,
       2, 1, 2, 0, 2, 0, 1, 2, 0, 2, 0, 1, 3, 1, 2, 3, 1, 3, 0, 2, 2, 1,
       3, 3, 0, 1, 1, 0, 0, 2, 0, 1, 1, 2, 3, 1, 0, 1, 1, 0, 0, 0, 3, 1,
       2, 3, 2, 2, 1, 1, 3, 0, 2, 2, 3, 0, 0, 3, 0, 2, 3, 2, 0, 3, 3, 0,
       3, 0, 3, 2, 3, 3, 1, 0, 1, 1, 0, 1, 2, 2, 3, 3, 0, 0, 0, 3, 0, 0,
       2, 3, 0, 1, 0, 2, 3, 2, 0, 3, 3, 0, 3, 3, 0, 2, 3, 3, 3, 2, 2, 2,
       1, 2, 0, 0, 0, 1, 0, 1, 0, 0, 3, 2, 3, 0, 3, 2, 2, 2, 3, 0, 3, 3,
       1, 0, 0, 3, 1, 0, 3, 2, 2, 0, 0, 0, 2, 2, 2, 2, 1, 0, 1, 3, 3, 1,
       1, 3, 1, 3, 1, 3, 0, 1, 1, 2, 2, 3, 1, 0, 1, 2, 3, 1, 2, 3, 1, 0,
       3, 3, 1, 2, 1, 1, 2, 3, 1, 1, 3, 2, 0, 3, 1, 0, 3, 3, 2, 3, 1, 2,
       3, 0, 0, 1, 0, 1, 3, 0, 0, 1, 3, 0, 3, 2, 1, 2, 1, 1, 0, 2, 0, 3,
       3, 1, 1, 0, 3, 2, 3, 1, 1, 2, 2, 3, 3, 1, 0, 2, 2, 2, 3, 0, 1, 1,
       3, 1, 0, 1, 1, 2, 3, 2, 1, 1, 2, 2, 1, 2, 0,

In [79]:
from collections import Counter
Counter(idade_discretizado)

Counter({1: 124, 0: 133, 2: 111, 3: 132})

#### Intervalos de Igual Frequência

- Neste tipo de discretização a ideia consiste em manter o mesmo número de valores (ou exemplos) em cada intervalo.

- Considerando $k$ o número de intervalos e $n$ o número de exemplos:
    1. Calculamos $N$, o número de exemplos por intervalo, como $N = \frac{n}{k}$
    - Calculamos o ranking dos valores da variável $\mathbf{x}$ 
    - Aplicamos o método de discretização de igual tamanho sobre o ranking dos elementos

In [87]:
k = 4
N = np.round(len(idade) / k)

from scipy.stats import rankdata
x = rankdata(idade)

L = (np.max(x) + np.min(x) - 1) / k
intervalos = (x - np.min(x)) // L

intervalos_igual_frequencia = intervalos.astype(int)
print(intervalos_igual_frequencia)

[1 0 2 0 2 0 3 2 0 0 3 2 0 1 2 0 0 3 0 1 1 0 2 2 3 3 3 3 3 0 0 0 2 2 0 2 0
 0 2 3 0 3 0 1 2 1 2 0 2 0 2 2 0 2 1 1 3 1 2 3 1 3 0 2 2 1 3 3 0 1 1 0 0 2
 0 1 1 2 3 1 0 1 1 0 0 0 3 1 2 3 2 2 1 1 3 0 2 2 3 0 0 3 0 2 3 2 0 3 3 0 2
 0 3 2 2 3 1 0 1 1 0 1 2 2 3 3 0 0 0 3 0 1 2 3 0 1 0 2 3 2 1 3 3 0 3 3 0 2
 3 3 3 2 2 2 1 2 0 0 0 1 0 1 0 0 3 2 3 0 3 2 2 2 3 0 3 3 1 0 0 3 1 0 3 2 2
 0 1 0 2 2 2 2 1 0 1 3 3 2 1 3 1 3 1 3 0 1 1 2 2 3 1 0 1 2 2 1 2 3 1 0 3 3
 1 2 1 1 2 3 1 1 3 2 0 3 1 0 3 3 2 3 1 2 3 1 0 1 0 1 3 0 0 1 3 0 3 2 1 2 1
 1 0 2 1 3 3 1 1 0 3 2 3 1 1 2 2 3 3 1 0 2 2 2 3 0 1 1 3 1 0 1 1 2 3 2 1 1
 2 2 1 2 0 3 0 0 0 3 2 0 2 0 0 2 0 1 0 2 3 3 1 2 3 2 0 3 0 1 2 3 2 3 1 1 1
 0 3 2 3 0 2 1 3 2 3 1 1 3 1 3 2 3 1 1 1 2 2 2 2 3 1 1 0 3 3 3 1 3 1 2 0 3
 2 0 2 0 1 3 1 2 2 0 1 0 0 2 3 3 2 0 1 2 3 3 3 3 0 1 0 1 3 2 1 1 3 0 3 3 1
 1 3 1 2 1 0 3 0 1 1 1 0 3 1 2 1 3 0 3 1 3 2 2 1 1 0 2 0 3 1 0 0 0 3 0 3 2
 1 2 1 3 2 0 0 3 2 1 0 1 1 2 3 2 1 0 3 0 1 0 2 0 1 0 1 1 1 0 2 3 2 0 0 1 0
 1 2 0 3 0 1 3 2 2 2 3 3 

In [88]:
from collections import Counter
Counter(intervalos)

Counter({1.0: 127, 0.0: 127, 2.0: 121, 3.0: 125})

In [89]:
df = pd.DataFrame({ "idade" : idade, "rank" : x, "intervalo" : intervalos_igual_frequencia} )
df

idade   rank  intervalo
0       38  200.5          1
1       13   60.5          0
2       73  360.0          2
3       10   43.5          0
4       76  372.0          2
..     ...    ...        ...
495     73  360.0          2
496     49  251.0          1
497     20   98.5          0
498     13   60.5          0
499     26  137.5          1

[500 rows x 3 columns]

## Transformação de Dados Qualitativos em Quantitativos

- Alguns algoritmos tem preferência por representações numéricas para atributos qualitativos, como as Redes Neurais.

- Os atributos qualitativos (ou categóricos) podem ser:
    - Ordinais: os seus valores possuem uma ordem, como avaliação: ruim, regular, bom e ótimo.
    - Nominais: os seus valores não possuem uma ordem, como cores, estado civil, e outros.

### Mapeamento Direto
- Para atributos nominais não é recomendado, pois pode inserir informações incorretas para o processo de aprendizado.
- Para os atributos ordinais é possível transformar os dados para números inteiros.
- Para isso, ordenamos os possíveis valores em ordem crescente, e atribuímos valores em ordem crescente aos possíveis valores, geralmente iniciando em 0 ou 1.
- Por exemplo, para a variável avaliação

|Valor | Mapeamento |
|------|------|
| ruim | 0 |
| regular | 1 |
| bom | 2 | 
| ótimo | 3 |

### Representação Binária

- Neste tipo de representação cada valor corresponde a uma sequência binária única.

- Em geral, usa-se a representação *One-Hot* que ativa um bit exclusivamente para cada valor da variável.

- Por exemplo, para o caso da variável ordinal *avaliação*:

|Valor | Mapeamento |
|------|------|
| ruim | 0001 |
| regular | 0010 |
| bom | 0100 | 
| ótimo | 1000 |

- Por exemplo, para a variável nominal *estado civil*:

|Valor | Mapeamento |
|------|------|
| casado | 00001 |
| solteiro | 00010 |
| viúvo | 00100 | 
| divorciado | 01000 |
| outro | 10000 |

- A representação binária é comumente utilizada para codificar variáveis em Redes Neurais.

### Outras Transformações de Dados

- Existem outras transformações dependentes da natureza do dado e da informação que queremos estudar.

- Por exemplo, se temos uma variável *data de nascimento*. Se estamos interessados em analisar o valor da idade, a data de nascimento não será suficiente para um método de aprendizado entender o que nós queremos.

In [ ]:
set.seed(1)
data.nasc <- sample(seq(as.Date('1950/01/01'), as.Date('2010/01/01'), by="day"), 20)
print(head(data.nasc))

In [ ]:
library(lubridate)
today <- Sys.Date()
age <- time_length( today - data.nasc, "years")
age <- floor(age) # Arredondamento pelo menor inteiro
head(data.table(data.nasc, idade = age))

- A data é um tipo de dado complexo, pois tem várias informações contidas nela.

- Os algoritmos de aprendizado também não irão saber que nela estão contidos: dia, mês e ano.

- E muito menos se o mês deve ser representado no formato textual, número inteiro ou representação binária. Portanto, nós devemos informar que tipo de informação queremos representar.

- Como já mencionado, os dados podem vir em qualquer formato, e como vão ser usados esses dados para visualização ou por algoritmo de aprendizado, vai depender de quem está realizando o processo de mineração.

- Então muitas vezes vamos precisar extrair todas essas informações do dado complexo

In [ ]:
dt <- data.table(data.nasc, 
                 dia.do.mes = mday(data.nasc), 
                 dia.da.semana = wday(data.nasc), 
                 dia.da.semana.extenso = wday(data.nasc, label = T), 
                 mes = month(data.nasc), 
                 mes.extenso = month(data.nasc, label = T), 
                 ano = year(data.nasc))
head(dt,5)

## Tratamento de Dados Desconhecidos

- Um problema comum em conjunto de dados é a presença de dados desconhecidos

- Geralmente sinalizados com: ?, NA, ou a ausência de valor mesmo

- Algumas abordagens comuns:
    - Eliminar o exemplo que possui o valor desconhecido
    - Substituir o valor de uma variável pelo seu valor esperado de acordo com sua distribuição. Geralmente média ou moda.
    - Verificar qual é o exemplo mais parecido no conjunto de dados, em relação aos outros atributos, e usar o valor do exemplo mais parecido


- Uma abordagem mais sofisticada é efetuar um aprendizado supervisionado sobre esse atributo para aprender sua relação com os outros atributos.
    - Modelo de classificação, quando o atributo com valor desconhecido é categórico.
    - Modelo de regressão, quando o atributo com valor desconhecido é numérico.
- Após, usa-se o modelo para predizer o valor desconhecido.
- Por exemplo, no conjunto de dados *iris*, vamos retirar 10 valores de *Sepal.Length*

In [90]:
set.seed(1)
desconhecidos <- sample( x = 1:nrow(iris), size = 10, replace = F)
print(desconhecidos)

SyntaxError: invalid syntax (<ipython-input-90-a7f7915e9917>, line 2)

In [91]:
my.iris <- data.table(iris)[,.(Sepal.Length,Sepal.Width,Petal.Length,Petal.Width)]
my.iris$Sepal.Length[desconhecidos] <- NA

SyntaxError: invalid syntax (<ipython-input-91-40e4c53c693c>, line 1)

- Construímos um modelo preditivo para o atributo *Sepal.Length*

In [ ]:
model <- lm(Sepal.Length ~ ., data = my.iris[-desconhecidos])

- Usamos esse modelo para predizer os valores desconhecidos de *Sepal.Length* 

In [ ]:
predicted <- predict(model, my.iris[desconhecidos,])

- Comparamos os valores preditos com os observados

In [ ]:
observed <- iris$Sepal.Length[desconhecidos]
results <- data.table(predicted, observed)
head(results,5)

- Verificamos se o nosso modelo foi melhor do que um modelo de substituição pela média

In [ ]:
mean.predictor <- mean(my.iris$Sepal.Length[-desconhecidos])

dt <- results[,.(observed, error.my.predictor = observed - predicted, error.mean.predictor = observed - mean.predictor)]

dt[,.(EMQ.my.predictor   = mean(error.my.predictor^2  ), 
      EMQ.mean.predictor = mean(error.mean.predictor^2) )]

- Isso indica que o Erro Médio Quadrático (EMQ) do preditor reduziu o erro da média de 0.72 para 0.15. 

- Podemos caluclar a redução de erro relativa:

$$\frac{0.72-0.15}{0.72} = 0.79$$

- E verificar uma redução de erro de $79\%$ usando o nosso modelo, em relação a usar apenas o valor da média

- Entre essas duas abordagens, a utilização de nosso modelo preditivo é preferível para preencher os dados desconhecidos no conjunto de dados, embora exista um EMQ de $0.15$.